src/lib/add-on/clients/add-on_auto.rb

```diff
@@ -1,328 +1,319 @@
+      AddOnProduct.modified
+    end
+
+    def modified
+      AddOnProduct.modified = true
+    end
+
+    def reset
+      AddOnProduct.add_on_products = []
+
+      {}
+    end
+
+    def change
+      Wizard.CreateDialog
+      AutoinstSoftware.pmInit
+      PackageCallbacks.InitPackageCallbacks
+
+      ret = RunAddOnMainDialog(
+        false,
+        true,
+        true,
+        Label.BackButton,
+        Label.OKButton,
+        Label.CancelButton,
+        false
+      )
+
+      UI.CloseDialog
+
+      ret
+    end
+
+    def export
+      AddOnProduct.Export
+    end
+
+    def write
+      AddOnProduct.add_on_products.each { |add_on| create_source(add_on) }
+
+      # reread agents, redraw wizard steps, etc.
+      AddOnProduct.ReIntegrateFromScratch
+
+      true
+    end
+
+    def read
+      if !PackageLock.Check
+        log.error("Cannot get package lock")
+
+        return false
+      end
+
+      log.info("Reseting Pkg")
+
+      Pkg.PkgApplReset
+      Pkg.PkgReset
+      Pkg.TargetInitialize(Installation.destdir)
+      Pkg.TargetLoad
+      Pkg.SourceStartManager(true)
+      Pkg.PkgSolve(true)
+
+      ReadFromSystem()
+    end
+
+  private
+
+    # Create repo and install product (if given)
+    #
+    # @param [Hash] add_on
+    def create_source(add_on)
+      media_url = media_url_for(add_on)
+      product = add_on.fetch("product", "")
+      product_dir = add_on.fetch("product_dir", "/")
+      retry_on_error = add_on.fetch("ask_on_error", false)
+
+      # Set addon specific sig-handling
+      AddOnProduct.SetSignatureCallbacks(product)
+
+      loop do
+        url = expand_url_for(add_on, media_url)
+        source_id = Pkg.SourceCreate(url, product_dir)
+
+        log.info("New source ID: #{source_id}")
+
+        if valid_source?(source_id)
+          retry_on_error = false
+          delete_and_halt(source_id) unless accepted_license?(add_on, source_id)
+
+          # bugzilla #260613
+          AddOnProduct.Integrate(source_id)
+
+          # FIXME: Why was outside the loop?
+          adjust_source_attributes(add_on, source_id)
+          install_product(product)
+        else
+          retry_on_error &&= retry_again?(product, media_url)
+          report_error_for(product, media_url) unless retry_on_error
         end
-        Builtins.y2milestone("Reseting Pkg")
-        Pkg.PkgApplReset
-        Pkg.PkgReset
 
-        Pkg.TargetInitialize(Installation.destdir)
-        Pkg.TargetLoad
-        Pkg.SourceStartManager(true)
-        Pkg.PkgSolve(true)
+        break unless retry_on_error
+      end
+    end
 
-        @ret = ReadFromSystem()
-      else
-        Builtins.y2error("unknown function: %1", @func)
-        @ret = false
+    # Verifies if source is valid
+    #
+    # @param [Integer|Nil] source_id
+    #
+    # @return [Boolean] true if given source id is valid; false otherwise
+    def valid_source?(source_id)
+      source_id != -1 && source_id != nil # rubocop:disable Style/NonNilCheck
```

Comment:
what about `![nil, -1].include?(source_id)`? Then you do not need rubocop disable.

Reply:
or better use constant like 
```
INVALID_SOURCE_IDS = [nil, -1]
....

!INVALID_SOURCE_IDS.include?(source_id)
```

Reply:
In fact, I was using your first suggestion but then I changed it for two main reasons:

   * It was not clear to me the use of nil value, as I said in https://github.com/yast/yast-add-on/pull/54#discussion_r197025382 (@lslezak already explained the reason why it must still there).
   * Despite to be less "ruby style", it seems to me more "simple" and readable the condition as it was.